### PR TITLE
[FIXED JENKINS-46673] Be able to use UDP ports as well

### PR DESCRIPTION
**Pull request by fbelzunc**
When using `ports={389, 3268}` in docker-fixtures it will only use the TCP protocol for those ports. It should be also possible to use UDP ports... i.e to use bind9 in AD plugin for DNS resolution we need to use the port 53 in both TCP and UDP.

@jglick @oleg-nenashev

### CODE REVIEW

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java [hunk 1]
```diff
@@ -223,5 +234,11 @@ private String getPortMapping(int port) {
                     : ipAddress + ":" + (portOffset + port) + ":" + port
             ;
         }
+
+        private String getUdpPortMapping(int UdpPort) {
+            return portOffset == null
+                    ? ipAddress + "::" + UdpPort
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java]: Do you not need `/udp` here too?

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java [hunk 2]
```diff
@@ -223,5 +234,11 @@ private String getPortMapping(int port) {
                     : ipAddress + ":" + (portOffset + port) + ":" + port
             ;
         }
+
+        private String getUdpPortMapping(int UdpPort) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java]: code style: `udpPort`

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java [hunk 3]
```diff
@@ -29,14 +29,24 @@
...
      */
     int[] ports() default {};
+
+    /**
+     * UDP ports that are exposed from this fixture.
+     *
+     * <p>
+     * When a container is started, these ports from the container are mapped to
+     * random ephemeral ports on the host. The actual ephemeral port number can
+     * be retried at runtime via {@link DockerContainer#port(int)}.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java]: `docker-inspect` offers no separate mapping for UDP ports?

Comment (fbelzunc) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java]: Correct, I did not know it. See the output:

```
pegaso:53310 fbelzunc$ docker port fc46392cb17f 53
0.0.0.0:53
pegaso:53310 fbelzunc$ docker port fc46392cb17f 53/udp
0.0.0.0:53
pegaso:53310 fbelzunc$ docker port fc46392cb17f
138/tcp -> 0.0.0.0:138
3268/tcp -> 0.0.0.0:3268
3269/tcp -> 0.0.0.0:3269
53/tcp -> 0.0.0.0:53
88/tcp -> 0.0.0.0:88
53/udp -> 0.0.0.0:53
135/tcp -> 0.0.0.0:135
139/tcp -> 0.0.0.0:139
389/tcp -> 0.0.0.0:389
445/tcp -> 0.0.0.0:445
464/tcp -> 0.0.0.0:464
```

**Comment by fbelzunc** [comment]
@jglick I think I correctly fixed everything and regarding

> Generally, cannot review such an API addition without a downstream PR using a timestamped snapshot dependency which demonstrates the new API being used successfully. (ci.jenkins.io supports Docker-aware agents so it should be possible to prove it with a public test pass.)

Sorry, I don't know where to start to achieve this. If there is any example, I can look at that and try.

**Comment by jglick** [comment]
> If there is any example, I can look at that and try.

There are a number of examples of using `docker-fixtures` in various plugins, as well as its original usage in `acceptance-test-harness`.

Anyway that is sort of missing the point, which is: you presumably filed this PR so you could use `docker-fixtures` in some place where you cannot currently use it, due to use of UDP-based protocols. And in order to see whether this PR actually works at all, you would need to actually develop some test specifying `udpPorts` on a fixture and calling `udpPort` in the test case. Preferably concurrently with this PR, so you are finding and fixing your own mistakes rather than relying on me to guess at them during code review. So, in which PR does that test live?

At any rate, once this and the downstream PR are actually ready for review, simply request me as a reviewer via the GitHub gesture and it will appear on my list.

**Comment by olivergondza** [comment]
I agree with Jesse, I hate to expose an API before we are certain it does work for requestor at least.

**Comment by fbelzunc** [comment]
@jglick Now, I see your point.

The downstream PR is https://github.com/jenkinsci/active-directory-plugin/pull/78 - I just requested you as a reviewer.

**Comment by fbelzunc** [comment]
@jglick I already created the downstream PR

* https://github.com/jenkinsci/active-directory-plugin/pull/78

I did the `clone/deploy` inside the build container where the ad-plugin builds.

https://github.com/fbelzunc/active-directory-plugin-build-container-with-docker-fixtures/blob/master/custom-dockerd-entrypoint.sh

File: pom.xml [hunk 4]
```diff
@@ -9,7 +9,7 @@
     </parent>
     <groupId>org.jenkins-ci.test</groupId>
     <artifactId>docker-fixtures</artifactId>
-    <version>1.3-SNAPSHOT</version>
+    <version>1.4-SNAPSHOT-fbelzunc-udp-ports</version>
```
Comment (jglick) [inline pom.xml]: No do not change the `version` in a PR. Just `mvn -DskipTests clean deploy` your branch.

File: Jenkinsfile [hunk 5]
```diff
@@ -12,11 +12,12 @@ pipeline {
     stages {
         stage('main') {
             steps {
-                sh 'mvn -B clean verify'
+                sh 'mvn -B clean package'
             }
             post {
                 success {
                     junit '**/target/surefire-reports/TEST-*.xml'
+                    archiveArtifacts 'target/docker-fixtures-1.4-SNAPSHOT.jar'
```
Comment (jglick) [inline Jenkinsfile]: revert

File: Jenkinsfile [hunk 6]
```diff
@@ -12,11 +12,12 @@ pipeline {
     stages {
         stage('main') {
             steps {
-                sh 'mvn -B clean verify'
+                sh 'mvn -B clean package'
```
Comment (jglick) [inline Jenkinsfile]: revert

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 7]
```diff
@@ -118,6 +118,25 @@ public int port(int n) {
...
+        assertRunning();
+        try {
+            if (sharingHostDockerService()) {
+                return n;
+            }
+            String out = Docker.cmd("port").add(cid, n + "/udp").popen().verifyOrDieWith("docker port command failed").trim();
+            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326"
+                throw new IllegalStateException(format("Udp port %d is not mapped for container %s", n, cid));
+
+            return Integer.parseInt(out.split(":")[1]);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: And then you are throwing away the first component. Surely you need a UDP equivalent of `ipBound`?

Comment (fbelzunc) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: @jglick I added the UDP equivalent for `ipBound`, but I don't need it. The reason why I don't need this is that the samba container is reachable inside the [build-docker-container](https://github.com/fbelzunc/active-directory-plugin-build-container-with-docker-fixtures) from the IP address `127.0.0.1` - all containers which share the same Docker Engine shared this behavior and it is likely not to change.

**Comment by jglick** [comment]
+0 I guess—from what I can see, the implementation would be appropriate if there were a use case for it, though I do not think the downstream PR should be using `docker-fixtures` at all.

**Pull request merged**